com.unity.shadergraph/Editor/ShaderFoundry/Utilities/ShaderUniformCollection.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace UnityEditor.ShaderFoundry
+{
+    internal class ShaderUniformCollection
+    {
+        Dictionary<string, UniformDeclarationData> visitedUniforms = new Dictionary<string, UniformDeclarationData>();
+        List<UniformDeclarationData> shaderUniforms = new List<UniformDeclarationData>();
+
+        public bool ReadOnly { get; private set; }
+        public IEnumerable<UniformDeclarationData> Uniforms => shaderUniforms;
+
+        public void Add(UniformDeclarationData uniform)
+        {
+            if (ReadOnly)
+            {
+                Debug.LogError("ERROR: attempting to add uniform to readonly collection");
+                return;
+            }
+
+            if (visitedUniforms.TryGetValue(uniform.Name, out var existingUniform))
+            {
+                ValidateAreEquivalent(uniform, existingUniform);
+                return;
+            }
+            visitedUniforms.Add(uniform.Name, uniform);
+            shaderUniforms.Add(uniform);
+        }
+
+        public void AddRange(IEnumerable<UniformDeclarationData> uniforms)
+        {
+            foreach (var uniform in uniforms)
+                Add(uniform);
+        }
+
+        public void Add(ShaderPropertyCollection shaderProperties)
+        {
+            foreach (var property in shaderProperties.Properties)
+            {
+                var propertyData = PropertyDeclarations.Extract(property.Type, property.Name, property.Attributes);
+                if (propertyData != null && propertyData.UniformDeclarations != null)
+                    AddRange(propertyData.UniformDeclarations);
+            }
+        }
+
+        public void SetReadOnly()
+        {
+            ReadOnly = true;
+        }
+
+        public bool HasDotsProperties()
+        {
+            foreach (var uniformDeclaration in Uniforms)
+            {
+                if (uniformDeclaration.DataSource == UniformDataSource.PerInstance)
```

Comment:
Is this always DOTS properties? Or can we declare instanced properties not for DOTS as well?

Reply:
We could add a bool field and update when adding a property if we're calling this at least once for each collection

Reply:
This is me effectively moving https://github.com/Unity-Technologies/Graphics/pull/7245/files#diff-7fc13037f08a23d2df289dfd63e80b84c6417a692784f482c55912e0c8109802R870 into the collection. This was previously mapping ShaderGraph's `HybridPerInstance` field which is synonymous with our `PerInstance` field. If we need to split it that'd be fine, but I think this covers things right now.

A fair amount could be cached in regards to properties (we rebuild them from the attributes a fair amount). I'll do a quick update for the bool here though.